inference-engine/src/mkldnn_plugin/nodes/cum_sum.cpp

```diff
@@ -120,75 +121,121 @@ class CumSumImpl: public ExtLayerBase {
+                if (reverse) {
+                    if (exclusive) {
+                        outputStart[offset*(shape[axis] - 1)] = 0;
+                        for (int64_t i = shape[axis] - 2; i >= 0; i--) {
+                            outputStart[i*offset] = inputStart[(i+1)*offset] + outputStart[(i+1)*offset];
+                        }
+                    } else {
+                        outputStart[offset*(shape[axis] - 1)] = inputStart[offset * (shape[axis] - 1)];
+                        for (int64_t i = shape[axis] - 2; i >= 0; i--) {
+                            outputStart[i*offset] = inputStart[i*offset] + outputStart[(i+1)*offset];
+                        }
                     }
                 } else {
-                    outputStart[0] = inputStart[0];
-                    for (size_t i = 1; i < shape5d[axis]; i++) {
-                        outputStart[i*offset] = inputStart[i*offset] + outputStart[(i-1)*offset];
+                    if (exclusive) {
+                        outputStart[0] = 0;
+                        for (size_t i = 1; i < shape[axis]; i++) {
+                            outputStart[i*offset] = inputStart[(i-1)*offset] + outputStart[(i-1)*offset];
+                        }
+                    } else {
+                        outputStart[0] = inputStart[0];
+                        for (size_t i = 1; i < shape[axis]; i++) {
+                            outputStart[i*offset] = inputStart[i*offset] + outputStart[(i-1)*offset];
+                        }
                     }
                 }
+
+                parallelItStep(counters, iterationRange);
             }
         });
     }
 
-    size_t getStartOffset(std::vector<size_t> &forStartOffset) {
-        return forStartOffset[N]*shape5d[C]*shape5d[D]*shape5d[H]*shape5d[W] + forStartOffset[C]*shape5d[D]*shape5d[H]*shape5d[W] +
-               forStartOffset[D]*shape5d[H]*shape5d[W] + forStartOffset[H]*shape5d[W] + forStartOffset[W];
+    void parallelItInit(size_t start, std::vector<size_t>& counters, std::vector<size_t>& iterationRange) {
+        auto itCounter = counters.rbegin();
+        auto itWork = iterationRange.rbegin();
+        while (itCounter != counters.rend()) {
+            *itCounter = start % *itWork;
+            start /= *itWork;
+            ++itCounter;
+            ++itWork;
+        }
     }
 
-    size_t getAxis(const Blob::CPtr& _axis, const Blob::CPtr& _data) {
+    void parallelItStep(std::vector<size_t>& counters, std::vector<size_t>& iterationRange) {
+        auto itCounter = counters.rbegin();
+        auto itWork = iterationRange.rbegin();
+
+        while (itCounter != counters.rend()) {
+            *itCounter = (*itCounter + 1) % *itWork;
+            if (*itCounter != 0) {
+                break;
+            }
+            ++itCounter;
+            ++itWork;
+        }
+    }
+
+    size_t getStartOffset(const std::vector<size_t> &forStartOffset, const std::vector<size_t>& strides) const {
```

Comment:
It makes sense to mark the function inplace since it is called in the `iwork` loop.  Same for `parallelItStep`